feat(migrations): Adds support for CommonModule to standalone migration

SkyZeroZx opened the pull request:
# Common to Standalone Migration

This migration replaces `CommonModule` imports with specific standalone imports from `@angular/common`.


## Usage

```bash
ng g @angular/core:common-to-standalone --path src/app/feature
```
  
## Example

**Before:**
```typescript
import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'app-example',
  standalone: true,
  imports: [CommonModule],
  template: `
    <div *ngIf="show">
      {{ data | async | json }}
    </div>
  `
})
export class ExampleComponent {
  show = true;
  data = Promise.resolve({ message: 'Hello' });
}
```

**After:**
```typescript
import { Component } from '@angular/core';
import { AsyncPipe, JsonPipe, NgIf } from '@angular/common';

@Component({
  selector: 'app-example',
  standalone: true,
  imports: [AsyncPipe, JsonPipe, NgIf],
  template: `
    <div *ngIf="show">
      {{ data | async | json }}
    </div>
  `
})
export class ExampleComponent {
  show = true;
  data = Promise.resolve({ message: 'Hello' });
}
``` 

## Benefits
- **Smaller bundle size**: Only imports needed directives and pipes instead of the entire CommonModule
- **Better tree-shaking**: Unused parts of @angular/common are eliminated


## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: https://github.com/angular/angular/issues/60808


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JeanMeche reviewed (commented):
Thanks for this. 
Can you make sure to include the corresponding docs in the change, see `adev/src/content/reference/migrations/overview.md`

JeanMeche commented on packages/core/schematics/test/common_to_standalone_migration_spec.ts:
```diff
@@ -0,0 +1,1809 @@
+
+        @Component({
+          selector: 'app-test',
+          standalone: true,
```
```suggestion
```
Can we remove the unnecessary `standalone: true` from the tests ?

JeanMeche replied:
We did have a migration in v19 that removed the `standalone: true` (based on a heuristic to make it idempotent)  but we didn't run it on the repo. 
Beside tests ensuring that `standalone: true` still works as expected, we should introduce new usages (au contraire, we should remove them as #63963 already did recently)

JeanMeche commented on packages/core/schematics/test/common_to_standalone_migration_spec.ts:
```diff
@@ -0,0 +1,1809 @@
+  });
+
+  describe('Dont migrate', () => {
+    it('when CommonModule dont import from @angular/common', async () => {
```
nit
```suggestion
    it('when CommonModule is not imported from @angular/common', async () => {
```

JeanMeche commented on packages/core/schematics/test/common_to_standalone_migration_spec.ts:
```diff
@@ -0,0 +1,1809 @@
+    });
+  });
+
+  describe('NgModule usage', () => {
```
I'm wondering if we should have this one behind a flag. There a lot of edge cases that might be hard to migrate (or get wrong). 

For example, what happens in case of a transitive import ? Can you add a test for it ?

SkyZeroZx replied:
I tried it with a TODO app (with edge case testing) and it didn't work as expected, I think it's an edge case, I'm not sure whether to add more logic to achieve this.

JeanMeche replied:
You're probably refering to the SCAM pattern (Signal Component Angular Module). It predates the existance of standalone. And indeed in that context it could make sense to migrate NgModules. 

But honestly, the initial requirement was made for standalone components only. If we start supporting NgModule we'll get reports that it either works only partially or that some edge cases are broken. 
Apps that uses the SCAM pattern should first run the Standalone migration then this one. 

I really think that it would help with the maintenability of the migrations to keep them as simple as necessary.

SkyZeroZx commented:
Updated addressing the feedback, the only remaining question is whether to support NgModules or only standalone components. If it’s possible to consult this in some way, that would be great.

JeanMeche commented on adev/src/content/reference/migrations/common-to-standalone.md:
```diff
@@ -0,0 +1,63 @@
+# Convert CommonModule usage to standalone imports
+
+This migration helps projects remove uses of `CommonModule` inside components and NgModules by adding the minimal set of directive and pipe imports each template requires (for example, `NgIf`, `NgFor`, `AsyncPipe`, etc.). It can also migrate NgModules by replacing `CommonModule` with explicit imports derived from their declared components' templates.
```
Can you update the description after your latest changes.

SkyZeroZx replied:
Updated, I forgot to remove it from the documentation.

JeanMeche commented on packages/core/schematics/migrations/common-to-standalone-migration/migration.ts:
```diff
@@ -0,0 +1,266 @@
+        if (nodeDecorators) {
+          const decorators = getAngularDecorators(typeChecker, nodeDecorators);
+
+          if (decorators.length > 0) {
+            const componentDecorator = decorators.find((d) => d.name === 'Component');
+
+            if (componentDecorator) {
```
can we do early returns instead, this will reduce the nesting of the function.

JeanMeche commented on packages/core/schematics/migrations/common-to-standalone-migration/migration.ts:
```diff
@@ -0,0 +1,266 @@
+    let isStandalone = false;
+
+    for (const property of config.properties) {
+      if (ts.isPropertyAssignment(property) && ts.isIdentifier(property.name)) {
```
nit: invert the condition to `continue` instead of nested.

JeanMeche commented on adev/src/content/reference/migrations/common-to-standalone.md:
```diff
@@ -0,0 +1,63 @@
+# Convert CommonModule usage to standalone imports
+
+This migration helps projects remove uses of `CommonModule` inside components by adding the minimal set of directive and pipe imports each template requires (for example, `NgIf`, `NgFor`, `AsyncPipe`, etc.). 
```
```suggestion
This migration helps projects remove imports of the `CommonModule` inside components by adding the minimal set of directive and pipe imports each template requires (for example, `NgIf`, `NgFor`, `AsyncPipe`, etc.). 
```

JeanMeche commented on adev/src/content/reference/migrations/overview.md:
```diff
@@ -33,7 +33,10 @@ Learn about how you can migrate your existing angular project to the latest feat
       Convert component templates to prefer style bindings over the `NgStyle` directives when possible.
+    </docs-card>
+  <docs-card title="CommonModule to standalone imports" link="Migrate now" href="reference/migrations/common-to-standalone">
+    Replace usages of `CommonModule` with the individual directives and pipes needed by templates when possible.
```
```suggestion
    Replace imports of the `CommonModule` with imports of the individual directives and pipes used in the templates when possible.
```

SkyZeroZx replied:
Done

JeanMeche commented on packages/core/schematics/migrations/common-to-standalone-migration/migration.ts:
```diff
@@ -0,0 +1,281 @@
+      }
+
+      if (templateVisitor.resolvedTemplates.length === 0) {
+        processResolvedTemplate(
```
That's when the module is imported but unused, right ? 
If yes, can you add a comment to mention that.

SkyZeroZx replied:
Yes, I will add a comment.

JeanMeche commented on packages/core/schematics/migrations/common-to-standalone-migration/migration.ts:
```diff
@@ -0,0 +1,281 @@
+  private hasCommonModuleInImports(decorator: ts.Decorator): boolean {
+    if (!ts.isCallExpression(decorator.expression)) return false;
+
+    const config = decorator.expression.arguments[0];
```
This is me thinking out loud: 

I think this is unsafe if we hit a `@Directive()`, ie with 0 arguments. iirc `ts.isObjectLiteralExpression(config)` will throw if `config` is undefined. 

But since we don't care about `Directive` decorator, we should only check component decorators. 

Either way, we need to make the check more robust.

SkyZeroZx replied:
Since we only process the Component decorator, this shouldn't happen. However, it could be "possible" that a user might have something strange with the decorator. I'm going to add a condition.

JeanMeche commented on packages/core/schematics/migrations/common-to-standalone-migration/migration.ts:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  private hasCommonModuleInArray(arrayLiteral: ts.ArrayLiteralExpression): boolean {
+    return arrayLiteral.elements.some(
+      (element) => ts.isIdentifier(element) && element.text === 'CommonModule',
+    );
```
This only works if the CommonModule isn't aliased, right ? 
If so, we should probably, handle the alias + add a test to cover that.

SkyZeroZx replied:
Updated I just added support for aliases with their respective tests

JeanMeche reviewed (commented):
Next round of comments. PTAL.

SkyZeroZx committed:
feat(migrations): Adds support for CommonModule to standalone migration

Introduces a migration that replaces CommonModule usage with individual imports from @angular/common, aligning with Angular's standalone component approach and improving module import clarity
